### DISCUSSION TITLE
MetaPetal / GPUImage3 - Implement Video Filters, More Image FIlters

### DISCUSSION THREAD

**Discussion by gintechsystems** [Yummypets/YPImagePicker#701] [Vote for Issue to faster implement.]
Hey guys,

I was thinking of implementing more filters along with the ability to have post processing filters for videos. Thoughts on this? Any reason not to do? I have heard in the past you cannot use instagram filters as they are copyrighted...not sure if possible or even true.

I would like to start adding more filters though outside of the default Apple provides. gintechsystems has done a few PRs to fix other problems in the past (audio / mp4 bugs) with this library and we continue to use it in all projects. It is the best picker out there that is open source!

@NikKovIos @s4cha 

Thanks
gintechsystems

**Comment by NikKovIos** [comment]
@gintechsystems thanks!) 
Any PR would be appreciated🙂

However filters should be customizable. User should have an option to enable/disable   any of them.

To make postprocessing videos the functional should be placed in one place. It needs refactoring. If you want, you can start doing that. Please use solid architecture, or just separate classes by logic.

And finally - make PRs as small as possible. You can make a lot of them.

**Reply by gintechsystems** [reply]
What do you mean filters should be customizable? I mean adding new filters in this library so when the user gets to the filters screen, they can choose additional ones I can add that are not already there.

Is post processing videos on the roadmap? I can get a head start. If you for some reason don't like the architecture I can always move things :)

Thanks for giving me the approval.

**Reply by NikKovIos** [reply]
I mean that programmer could choose what filters to use. Like now.
No, processing not in roadmap) You can move, BUT in small PRs, not in huge one)

**Reply by gintechsystems** [reply]
I understand programmers could, but only what is available or if they create their own YPFilter. What if they don't know how to create one? Know what I mean. I think we should have more to "choose from" from just installing the library.

okay sounds good, will do what I can.

**Reply by NikKovIos** [reply]
" I think we should have more to "choose from" from just installing the library"
Yes.

**Reply by gintechsystems** [reply]
Great, maybe for starters we can add customization to allow users to choose what is visible or not (enable / disable).

**Reply by NikKovIos** [reply]
The way it's implemented with photo filter, I think, is flexible.

**Reply by gintechsystems** [reply]
I will double check and let you know my thoughts, its been awhile since I looked at this.

**Comment by NikKovIos** [comment]
In app you can use custom filters.

```swift
public typealias FilterApplierType = ((_ image: CIImage) -> CIImage?)

public struct YPFilter {
    var name = ""
    var applier: FilterApplierType?
    
    public init(name: String, coreImageFilterName: String) {
        self.name = name
        self.applier = YPFilter.coreImageFilter(name: coreImageFilterName)
    }
    
    public init(name: String, applier: FilterApplierType?) {
        self.name = name
        self.applier = applier
    }
}
```

**Reply by gintechsystems** [reply]
I am not sure how this relates to the original thread?
